### PR TITLE
WindowsPB : Add Visual Studio 2022 installation

### LINKED ISSUE #3168: Ansible request for Visual Studio 2022 on Windows x64 & Arm64
Include a Visual Studio 2022 Install from a predefined layout in the windows playbook.

- Request for new playbook addition

Details:

As part of the alignment of compilers, we require a VS2022 install on windows x64.

### PR DESCRIPTION
**Pull request by steelhead31**
Fixes #3168 

Install VS2022 on windows 10 & 11, and Windows Server 2016, 2019 & 2022, either from a pre-defined layout or from the community download in a similar fashion to VS2019.

VS2022 is now a 64-bit application.

##### Checklist

- [X] commit message has one of the [standard prefixes](https://github.com/adoptium/infrastructure/blob/master/CONTRIBUTING.md#commit-messages)
- [X] playbook changes run through [VPC](https://ci.adoptium.net/view/Tooling/job/VagrantPlaybookCheck/) or [QPC](https://ci.adoptium.net/view/Tooling/job/QEMUPlaybookCheck/) (if you have access)

VPC passed: https://ci.adoptium.net/job/VagrantPlaybookCheck/1725/

### CODE REVIEW

**Comment by gdams** [comment]
@steelhead31 I believe that VS2022 has a Native arm64 version. We probably want to make sure that the correct version gets installed on our arm64 hosts if possible.

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/MSVS_2022/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  when: (not vs2022_installed.stat.exists)
+  tags: adoptopenjdk
+#
+## Check If VS2019 Has Been Installed From Layout Even If Not AdoptOpenJDK
+- name: Test if VS 2022 is installed
+  win_stat:
+    path: 'C:\Program Files\Microsoft Visual Studio\2022\Community'
+  register: vs2022_installed
+
+# Download & Install VS2019 When No Layout & Not AdoptOpenJDK
```
Comment (gdams) [inline ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/MSVS_2022/tasks/main.yml:87]:
```suggestion
# Download & Install VS2022 When No Layout & Not AdoptOpenJDK
```

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/MSVS_2022/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+# Exit Play If No VS2022 Layout & AdoptOpenJDK = true
+# This Is Defined Behaviour - This Shouldnt Occur But If It Does, Its Captured Here
+- name: Exit With Error When No Layout Within AdoptOpenJDK
+  fail:
+    msg: "The VS2022 Layout File Appears To Be Missing From Vendor_Files"
+  when: (not vs2022_layout_ready.stat.exists) and (not vs2022_installed.stat.exists)
+  tags: adoptopenjdk
+
+# Install VS2019 From Layout For AdoptOpenJDK
```
Comment (gdams) [inline ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/MSVS_2022/tasks/main.yml:51]:
```suggestion
# Install VS2022 From Layout For AdoptOpenJDK
```

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/MSVS_2022/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  tags: adoptopenjdk
+#
+## Check If VS2019 Has Been Installed From Layout Even If Not AdoptOpenJDK
+- name: Test if VS 2022 is installed
+  win_stat:
+    path: 'C:\Program Files\Microsoft Visual Studio\2022\Community'
+  register: vs2022_installed
+
+# Download & Install VS2022 When No Layout & Not AdoptOpenJDK
+# This is the target that you're redirected to when you go to https://aka.ms/vs/16/release/vs_community.exe
```
Comment (gdams) [inline ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/MSVS_2022/tasks/main.yml:88]:
```suggestion
# This is the target that you're redirected to when you go to https://aka.ms/vs/17/release/vs_community.exe
```

**Comment by andrew-m-leonard** [comment]
> > @steelhead31 I believe that VS2022 has a Native arm64 version. We probably want to make sure that the correct version gets installed on our arm64 hosts if possible. (This might already be happening but thought it was worth checking)
> 
> There is a single installer for both Visual Studio x64 and Arm64 architectures. The Visual Studio 2022 installer detects if the system architecture is Arm64 and if so, it'll download and install the Arm64 version of Visual Studio on your Arm64 device (support is for Windows 11). If you are installing the product via a layout, you'll need to [explicitly configure the layout to include ARM binaries](https://learn.microsoft.com/en-us/visualstudio/install/use-command-line-parameters-to-install-visual-studio?view=vs-2022#layout-command-and-command-line-parameters).
> 
> So it looks like we need to configure our layout file to support these files if we haven't already

@gdams We created the install layotu using the following:
```
vs_community.exe --layout VSLayout2022 --add Microsoft.VisualStudio.Workload.NativeDesktop;includeRecommended;includeOptional --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64 --add Microsoft.VisualStudio.Component.VC.Tools.ARM64 --add Microsoft.VisualStudio.Component.VC.ATL.ARM64 --add Microsoft.VisualStudio.Component.VC.MFC.ARM64
```

**Comment by gdams** [comment]
> > > @steelhead31 I believe that VS2022 has a Native arm64 version. We probably want to make sure that the correct version gets installed on our arm64 hosts if possible. (This might already be happening but thought it was worth checking)
> > 
> > 
> > There is a single installer for both Visual Studio x64 and Arm64 architectures. The Visual Studio 2022 installer detects if the system architecture is Arm64 and if so, it'll download and install the Arm64 version of Visual Studio on your Arm64 device (support is for Windows 11). If you are installing the product via a layout, you'll need to [explicitly configure the layout to include ARM binaries](https://learn.microsoft.com/en-us/visualstudio/install/use-command-line-parameters-to-install-visual-studio?view=vs-2022#layout-command-and-command-line-parameters).
> > So it looks like we need to configure our layout file to support these files if we haven't already
> 
> @gdams We created the install layout using the following:
> 
> ```
> vs_community.exe --layout VSLayout2022 --add Microsoft.VisualStudio.Workload.NativeDesktop;includeRecommended;includeOptional --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64 --add Microsoft.VisualStudio.Component.VC.Tools.ARM64 --add Microsoft.VisualStudio.Component.VC.ATL.ARM64 --add Microsoft.VisualStudio.Component.VC.MFC.ARM64
> ```

@andrew-m-leonard can you please add `--arch all` to that command as that will generate an x64 only Layout file.

**Comment by andrew-m-leonard** [comment]
My installation on a x86_86 Win2022, shows the "arm" runtimes present
```
 Directory of C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\redist\arm\Microsoft.VC140.CRT

08/24/2023  08:25 AM    <DIR>          .
08/24/2023  08:25 AM    <DIR>          ..
05/30/2023  04:12 PM           275,856 concrt140.dll
05/30/2023  04:12 PM           521,584 msvcp140.dll
05/30/2023  04:12 PM           328,080 vccorlib140.dll
05/30/2023  04:12 PM            72,080 vcruntime140.dll
```

**Comment by steelhead31** [comment]
Latest VPC in progress: https://ci.adoptium.net/job/VagrantPlaybookCheck/1717/

**Comment by steelhead31** [comment]
Requires a fix for vs2022 dll registration.

**Comment by steelhead31** [comment]
VPC is currently failing on windows 2012, as VS2022 isn't supported, I'll update the playbook to only install when its only on a supported version.

**Review by andrew-m-leonard — APPROVED** [review]
looks good

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/main.yml [hunk 4]
```diff
@@ -73,6 +73,8 @@
       when: ansible_architecture == "64-bit"
     - role: MSVS_2019                   # OpenJ9
       tags: MSVS_2019
+    - role: MSVS_2022                   # OpenJ9
```
Comment (gdams) [inline ansible/playbooks/AdoptOpenJDK_Windows_Playbook/main.yml:76]:
```suggestion
    - role: MSVS_2022
```

**Review by gdams — APPROVED** [review]

**Pull request merged**